File: paddle/fluid/framework/details/print_utils.cc
// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <sstream>  // std::stringstream
#include <string>

#include "paddle/fluid/framework/details/print_utils.h"
#include "paddle/fluid/framework/executor.h"
#include "paddle/fluid/framework/lod_tensor.h"
#include "paddle/fluid/framework/op_registry.h"
#include "paddle/fluid/framework/var_type.h"
#include "paddle/fluid/framework/variable.h"

namespace paddle {
namespace framework {
namespace details {

std::string GetVariableReadableData(const framework::Tensor* input_tensor,
                                    const std::string& message,
                                    const std::string& name) {
  std::stringstream ss;
  auto& in_tensor = *input_tensor;
  if (!in_tensor.IsInitialized()) {
    ss << message << " " << name << " onit initialized";
    return ss.str();
  }

  auto dtype = in_tensor.type();
  size_t size = in_tensor.numel();
  size_t summarize = 10;

  framework::Tensor tensor;
  // tensor.set_lod(tensor.lod());
  tensor.Resize(tensor.dims());

  if (platform::is_cpu_place(in_tensor.place())) {
    tensor.ShareDataWith(in_tensor);
  } else {
    // copy data to cpu to print
    platform::CPUPlace cpu_place;
    framework::TensorCopy(in_tensor, cpu_place, &tensor);
  }
  void* data = const_cast<void*>(tensor.data<void>());

  ss << message << " " << name << ":";
  if (IsType<const float>(dtype)) {
    ss << GetArrayReadableData<float>(0, 1, size, summarize, data);
  } else if (IsType<const double>(dtype)) {
    ss << GetArrayReadableData<double>(0, 1, size, summarize, data);
  } else if (IsType<const int>(dtype)) {
    ss << GetArrayReadableData<int>(0, 1, size, summarize, data);
  } else if (IsType<const int64_t>(dtype)) {
    ss << GetArrayReadableData<int64_t>(0, 1, size, summarize, data);
  } else if (IsType<const bool>(dtype)) {
    ss << GetArrayReadableData<bool>(0, 1, size, summarize, data);
  } else {
    // TODO(gongwb): add more data types support.
    ss << "\tdata: unprintable type: " << dtype.name();
  }

  return ss.str();
}

std::string GetVariableReadableData(const Scope& scope,
                                    const std::string& message,
                                    const std::string& name) {
  std::stringstream ss;

  auto var = scope.FindVar(name);
  if (var == NULL) {
    ss << "can't find " << message << " " << name;
    return ss.str();
  }

  if (var->IsType<framework::SelectedRows>()) {
    return GetVariableReadableData(
        &var->GetMutable<framework::SelectedRows>()->value(), message, name);
  } else if (var->IsType<framework::LoDTensor>()) {
    return GetVariableReadableData(var->GetMutable<framework::LoDTensor>(),
                                   message, name);
  } else {
    PADDLE_ENFORCE(false, "only support SelectedRows and lodTensor");
  }

  return "";
}
}  // namespace details
}  // namespace framework
}  // namespace paddle

File: python/paddle/utils/print_utils.py
# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import functools


def print_args(func):
    @functools.wraps(func)
    def inner(*args, **kwargs):
        function_name = func.__name__
        from paddle.jit.dy2static.utils import parse_arg_and_kwargs

        api_params, api_defaults = parse_arg_and_kwargs(func)
        import collections

        inputs = collections.OrderedDict()
        params = collections.OrderedDict()
        from paddle import Tensor

        # from paddle import Tensor
        for i in range(len(args)):
            if isinstance(args[i], Tensor):
                inputs[api_params[i]] = args[i]
            elif (
                isinstance(args[i], (list, tuple))
                and len(args[i]) > 0
                and isinstance(args[i][0], Tensor)
            ):
                inputs[api_params[i]] = args[i]
            else:
                params[api_params[i]] = args[i]
        for key, value in kwargs.items():
            if type(value) == Tensor:
                inputs[key] = value
            else:
                params[key] = value
        log_msg = f"{{function_name : {function_name}, "
        log_msg += "inputs: { "
        for name, value in inputs.items():
            shape = []
            if (
                isinstance(value, (list, tuple))
                and len(value) > 0
                and isinstance(value[0], Tensor)
            ):
                for v in value:
                    shape.append(v.shape)
                input_type = "List(Tensor)"
            else:
                shape = value.shape
                input_type = str(type(value))
            log_msg += f"{{ {name}, type: {input_type}, shape: {shape} }}, "
        log_msg += "}, "
        log_msg += "params: [ "
        for name, value in params.items():
            log_msg += f"{name}: {str(value)}, "
        log_msg += "]}"
        print(log_msg, flush=True)
        return func(*args, **kwargs)

    return inner